helix-core/src/main/java/org/apache/helix/task/JobConfig.java

```diff
@@ -147,10 +147,17 @@
     StartTime,
 
     /**
-     * The expiration time for the job
+     * The expiration time for the job if it's completed; once the expiry is reached and the job is
+     * completed, the job will be purged
      */
     Expiry,
 
+    /**
+     * The expiration time for the job if it's failed or timed out; once the expiry is reached and
+     * the job has failed or timed out, the job will be purged
+     */
+    TerminalStateExpiry,
```

Comment:
This may not be a good idea for that. It means user has to add a new config for each their job. And the existing running jobs and tasks are not benefiting from that.

If we really want to support that, my suggestion is to have a global TASK config similar to cluster config. Let them enable for failed/timeout job purge. That would be much easier for user.

Reply:
The current `expiry` config we're using is on per-job level, though.

Reply:
I think it makes more sense to keep this config in the as job config because we are treating this config similar to Expiry. If anybody want to use it, they can add this config.

Reply:
Then there could be too many expiry to configured by user. Even if we want to support this at per job level, we need a boolean flag to show whether we should clean them after expiry when it fails or timed out. From time perspective, I dont think user has any difference with time number.

But from user migration perspective, adding to job level is very cost and roll out.

Reply:
> The current `expiry` config we're using is on per-job level, though. I imagine there's a reason behind that?

At job level is make sense that different job could have different time to persist for further investigation.

@alirezazamani And we dont have to make the behavior to fine-grained at  job level. Because we defined the default behavior not purge failed/timeout jobs gloabally. That's why I am suggesting not adding the options at job level.

Reply:
@dasahcc The default behavior being not purging is already implemented: not specifying the expiry = not purging; we don't need an extra boolean flag. 

I understand where your came from, but the flexibility should be a good thing; after all, wouldn't users be more inclined to investigate failed/timed-out jobs comparing to completed jobs? The investigation flexibility should be more suited for the terminal state expiry.

Reply:
@NealSun96 Even we dont setup the expiry, there is a default value for expiry is one day. That's why I am saying the default behavior is cleaning up successful job. That's our assumption.

We need to balance the operability and flexibility. If you really consider flexibility, we should have cluster level config + job level. Then the logic will be much complicated.

My suggestion is still having a global change flag instead of to fine-grained, because we dont see any fine-grained control request from any of our customers. And from operability perspective, it will be very hard for them roll out since it requires code level change and redeploy all their scheduler code. But if we support it from cluster level, it they can dynamically change it with our latest controller deployment. That will be much simpler.

Reply:
Resolved offline.